File: ProjectRSC/Server/scripts/python/skills/Mining.py
from org.darkquest.gs.plugins.listeners.action import ObjectActionListener
from org.darkquest.gs.plugins.listeners.executive import ObjectActionExecutiveListener
from org.darkquest.gs.plugins import PlugInterface
from org.darkquest.gs.external import EntityHandler, GameObjectLoc, ObjectMiningDef, ItemDef
from org.darkquest.gs.model import Entity, Point, GameObject
from org.darkquest.config import Constants, Formulae
from org.darkquest.gs.world import World
from org.darkquest.gs.tools import DataConversions

'''
@author: GORF
Mining skill
'''

class Mining(PlugInterface, ObjectActionListener, ObjectActionExecutiveListener):
	
	VALID_ROCKS = [176, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 195, 196, 210, 211]
	VALID_PICKAXES = [1262, 1261, 1260, 1259, 1258, 156]
	PICKAXE_LEVELS = [41, 31, 21, 6, 1, 1]
	
	REQ_USE_LEVELS = {1262:41, 1261:31, 1260:21, 1259:6, 1258:1, 156:1}
	RETRIES = {1262:12, 1261:8, 1260:5, 1259:3, 1258:2, 156:1}
	BONUS = {1262:12, 1261:10, 1260:8, 1259:6, 1258:4, 156:2}
	
	def onObjectAction(self, gameObject, command, player):
		script = player.getScriptHelper()
		mining_def = EntityHandler.getObjectMiningDef(gameObject.getID())
		cur_level = script.getCurrentLevel(player.SkillType.MINING)
		
		if player.isBusy() or not player.withinRange(gameObject, 1):
			return
		
		if mining_def == None or mining_def.getRespawnTime() < 1:
			script.displayMessage("There is currently no ore available in this rock.")
			return
		
		script.occupy()
		
		if player.click == 1:
			prospected_ore = script.getItem(mining_def.getOreId())
			script.displayMessage("You examine the rock for ores")
			script.sleep(2000)
			script.displayMessage("This rock contains " + str(prospected_ore.getDef().getName()) + ".")
			script.release()
			return
		
		axe_id = self.determineAxe(cur_level, script)
		
		if axe_id < 0:
			script.displayMessage("You need a pickaxe to mine this rock.")
			script.release()
			return
		
		retries = self.determineAccuracy(axe_id)
		req_use_level = self.determineReqUseLevel(axe_id)
		req_ore_level = mining_def.getReqLevel()
		
		if req_use_level > cur_level: # PICKAXE REQ LEVEL CHECK
			script.displayMessage("You need to be level " + str(req_use_level) + " to use this pick.")
			script.release()
			return
		elif cur_level < req_ore_level: # ORE REQ LEVEL CHECK
			script.displayMessage("You need to be level " + str(req_ore_level) + " to mine this ore.")
			script.release()
			return
		elif player.getFatigue() >= 7500: # FATIGUE CHECK
			script.displayMessage("You are too tired to mine this rock")
			script.release()
			return

		self.handleMining(gameObject, script, mining_def, axe_id, retries)
		script.release()
	
	def determineAxe(self, cur_level, script):
		for axe in self.VALID_PICKAXES:
			for level in self.PICKAXE_LEVELS:
				if script.hasItem(axe):
					if cur_level >= level:
						return axe
		return -1
	
	def determineReqUseLevel(self, axe_id):
		if self.REQ_USE_LEVELS.has_key(axe_id):
			return self.REQ_USE_LEVELS[axe_id]
		return -1
	
	def determineAccuracy(self, axe_id):
		if self.RETRIES.has_key(axe_id):
			return self.RETRIES[axe_id]
		return -1
	
	def determineRespawn(self, mining_def, script):
		respawn = mining_def.getRespawnTime() * 1000
		total_players = script.getTotalPlayers()
		
		#if mining_def.getReqLevel() == 20 or mining_def.getReqLevel() == 40: # SILVER/GOLD
		#	respawn = (2 - (total_players / 2000)) * 60000
		#elif mining_def.getReqLevel() == 30: # COAL
		#	respawn = (60 - ((3 * total_players) / 200)) * 1000
		#elif mining_def.getReqLevel() == 55: # MITH
		#	respawn = (4 - (total_players / 1000)) * 60000
		#elif mining_def.getReqLevel() == 70: # ADDY
		#	respawn = (8 - (total_players / 500)) * 60000
		#elif mining_def.getReqLevel() == 85: # RUNE
		#	respawn = (25 - (total_players / 160)) * 60000
		
		return respawn
	
	def getOre(self, script, mining_def, mining_level, axe_id):
		 level_diff = mining_level - mining_def.getReqLevel()
		 
		 if level_diff > 50:
			 return script.getRandom(0, 9) != 1
			
		 if level_diff < 0:
			 return False
		 
		 bonus = self.BONUS[axe_id]
		 level_diff = level_diff + bonus
		 if level_diff > 40:
		 	level_diff = 60
		 else:
		 	level_diff = 20 + level_diff
		 return DataConversions.percentChance(level_diff)
	
	def handleMining(self, game_object, script, mining_def, axe_id, retries):
		player = script.getPlayer()
		if player.lastMineTries == -1: 
			player.lastMineTries = 0
		
		player.lastMineTries = player.lastMineTries + 1
		script.sendSound("mine")
		script.showBubble(1258)
		script.displayMessage("You swing your pick at the rock...")
		self.miningEvent(game_object, script, mining_def, axe_id, retries)
	
	def miningEvent(self, game_object, script, mining_def, axe_id, retries):
		player = script.getPlayer()
		script.sleep(1500)
		ore = script.getItem(mining_def.getOreId())
		
		if self.getOre(script, mining_def, script.getCurrentLevel(player.SkillType.MINING), axe_id):
			if script.getRandom(0, 200) == 0: # FOUND GEM
				script.addItem(Formulae.getGem(), 1)
				script.advanceStat(player.SkillType.MINING, 100, True)
				script.displayMessage("You found a gem!")
				player.lastMineTries = -1
			else: # NORMAL ORE
				script.addItem(mining_def.getOreId(), 1)
				script.displayMessage("You manage to obtain some " + ore.getDef().getName() + ".")
				script.advanceStat(player.SkillType.MINING, mining_def.getExp(), True)
				player.lastMineTries = -1
				respawn = self.determineRespawn(mining_def, script)
				new_rock = script.createNewObject(game_object)
				script.spawnObject(game_object.getLocation(), 98, game_object.getDirection(), game_object.getType(), True, new_rock.getLoc(), respawn)
				if not player.getInventory().full() and Constants.GameServer.BATCH_EVENTS:
					script.sleep(500)
					self.handleMining(ore, script)
		else:
			script.displayMessage("You only succeed in scratching the rock.")
			player.isMining(False)
			script.sleep(500)
			last_tries = player.lastMineTries
			if last_tries < retries:
				self.handleMining(game_object, script, mining_def, axe_id, retries)
		script.release()
    
	def blockObjectAction(self, gameObject, command, player):
		return command == "mine" or command == "prospect" and player.click == 1

File: ProjectRSC/Server/src/org/darkquest/gs/plugins/skills/Mining.java
package org.darkquest.gs.plugins.skills;

import org.darkquest.config.Constants;
import org.darkquest.config.Formulae;
import org.darkquest.gs.event.ShortEvent;
import org.darkquest.gs.event.SingleEvent;
import org.darkquest.gs.external.EntityHandler;
import org.darkquest.gs.external.ObjectMiningDef;
import org.darkquest.gs.model.*;
import org.darkquest.gs.plugins.listeners.action.ObjectActionListener;
import org.darkquest.gs.plugins.listeners.executive.ObjectActionExecutiveListener;
import org.darkquest.gs.tools.DataConversions;
import org.darkquest.gs.world.World;

import java.util.Arrays;

public final class Mining implements ObjectActionListener, ObjectActionExecutiveListener {

	static int[] ids;

	static {
		ids = new int[]{176, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 195, 196, 210, 211};
		Arrays.sort(ids);
	}

	@Override
	public void onObjectAction(final GameObject object, String command, Player owner) {
		/*if (Arrays.binarySearch(ids, object.getID()) >= 0) {
			handleMining(object, owner, owner.click);
			return;
		} */
	}

	public int getAxe(Player p) {
		int lvl = p.getCurStat(14);

		for (int i = 0; i < Formulae.miningAxeIDs.length; i++) {
			if (p.getInventory().countId(Formulae.miningAxeIDs[i]) > 0) {
				if (lvl >= Formulae.miningAxeLvls[i]) {
					return Formulae.miningAxeIDs[i];
				}
			}
		} 
		return -1; 
	} 

	public void handleMining(final GameObject object, Player owner, int click) {
		if (owner.isBusy()) {
			return;
		}
		
		if (!owner.withinRange(object, 1)) {
			return;
		}
		
		final GameObject newobject = World.getWorld().getTile(object.getX(), object.getY()).getGameObject(); //???
		final ObjectMiningDef def = EntityHandler.getObjectMiningDef(newobject.getID());
		
		if (def == null || def.getRespawnTime() < 1) {
			owner.getActionSender().sendMessage("There is currently no ore available in this rock.");
			return;
		}
		
		final InvItem ore = new InvItem(def.getOreId());
		
		if (owner.click == 1) {
			owner.setBusy(true);
			owner.getActionSender().sendMessage("You examine the rock for ores");
			World.getWorld().getDelayedEventHandler().add(new SingleEvent(owner, 2000) {
				@Override
				public void action() {
					owner.getActionSender().sendMessage("This rock contains " + ore.getDef().getName() + ".");
					owner.setBusy(false);
				}
			});
			return;
		}

		final int axeId = getAxe(owner);

		if (axeId < 0) {
			owner.getActionSender().sendMessage("You need a pickaxe to mine this rock.");
			return;
		}

		int retrytimes = -1;
		final int mineLvl = owner.getCurStat(14);
		int reqlvl = 1;
		switch (axeId) {
		case 1258:
			retrytimes = 2;
			break;
		case 1259:
			retrytimes = 3;
			reqlvl = 6;
			break;
		case 1260:
			retrytimes = 5;
			reqlvl = 21;
			break;
		case 1261:
			retrytimes = 8;
			reqlvl = 31;
			break;
		case 1262:
			retrytimes = 12;
			reqlvl = 41;
			break;
		}

		final int shots = retrytimes;

		if (reqlvl > mineLvl) {
			owner.getActionSender().sendMessage("You need to be level " + reqlvl + " to use this pick.");
			return;
		}
		if(mineLvl < def.getReqLevel()) { // mine ore check
			owner.getActionSender().sendMessage("You need to be level " + def.getReqLevel() + " to mine this ore.");
			return;
		}
		if(owner.getFatigue() >= 7500) {
			owner.getActionSender().sendMessage("You are too tired to mine this rock");
			return;
		}
		owner.setBusy(true);
		if(owner.lastMineTries == -1) {
			owner.lastMineTries = 0;
		}
		owner.lastMineTries++;
		owner.getActionSender().sendSound("mine");
		Bubble bubble = new Bubble(owner, 1258); // default back to iron only (verified http://www.youtube.com/watch?v=CZtVBarTMjY)
		
		for (Player p : owner.getViewArea().getPlayersInView()) {
			p.informOfBubble(bubble);
		}

		owner.getActionSender().sendMessage("You swing your pick at the rock...");
		World.getWorld().getDelayedEventHandler().add(new ShortEvent(owner) {
			public void action() {
				if (owner.getCurStat(14) < def.getReqLevel()) { //updated to say you only scratched it - hik
					owner.getActionSender().sendMessage("You only succeed in scratching the rock.");
					owner.lastMineTries = -1;
				} else if (Formulae.getOre(def, owner.getCurStat(14), axeId)) {
					if (DataConversions.random(0, 200) == 0) {
						InvItem gem = new InvItem(Formulae.getGem(), 1);
						owner.incExp(14, 100, true);
						owner.getInventory().add(gem);
						owner.getActionSender().sendMessage("You found a gem!");
						owner.lastMineTries = -1;
					} else {
						owner.getInventory().add(ore);
						owner.getActionSender().sendMessage("You manage to obtain some " + ore.getDef().getName() + ".");
						owner.incExp(14, def.getExp(), true);
						owner.getActionSender().sendStat(14);
						owner.lastMineTries = -1;
						world.registerGameObject(new GameObject(object.getLocation(), 98, object.getDirection(), object.getType()));

						int respawn = def.getRespawnTime() * 1000;
						
						//formulae for silver, coal, gold, mith, addy, and rune rocks for respawn times
						/*if (def.getReqLevel() == 20) { //silver
							respawn = (2 - (world.countPlayers() / 2000)) * 60000; //same as gold
						} else if (def.getReqLevel() == 30) { //coal
							respawn = (60 - ((3 * world.countPlayers()) / 200)) * 1000; //this is already in seconds
						} else if (def.getReqLevel() == 40) { //gold
							respawn = (2 - (world.countPlayers() / 2000)) * 60000; //same as silver
						} else if (def.getReqLevel() == 55) { //mith
							respawn = (4 - (world.countPlayers() / 1000)) * 60000;
						} else if (def.getReqLevel() == 70) { //addy
							respawn = (8 - (world.countPlayers() / 500)) * 60000;
						} else if (def.getReqLevel() == 85) { //rune
							respawn = (25 - (world.countPlayers() / 160)) * 60000;
						} */
						world.delayedSpawnObject(newobject.getLoc(), respawn);						
					}
					owner.getActionSender().sendInventory();
					
					if (!owner.getInventory().full() && Constants.GameServer.BATCH_EVENTS) {
						world.getDelayedEventHandler().add(new SingleEvent(owner, 500) {
							public void action() {
								handleMining(object, owner, owner.click);
							}
						});
					}
				} else {
					owner.getActionSender().sendMessage("You only succeed in scratching the rock.");
					owner.isMining(false);
					world.getDelayedEventHandler().add(new SingleEvent(owner, 500) {
						public void action() {
							if(owner.lastMineTries < shots) {
								handleMining(object, owner, owner.click);
							}
						}
					});

				}
				owner.setBusy(false);
			}
		});
	}

	@Override
	public boolean blockObjectAction(GameObject obj, String command,
			Player player) {
		if (Arrays.binarySearch(ids, obj.getID()) >= 0) {
			handleMining(obj, player, player.click);
			return true;
		}
		if(command.equals("mine") || command.equals("prospect")) {
			return true;
		}
		return false;
	}
}